### ISSUE TITLE
NO_CONVERGENCE shows in one iteration with dynamic autodiff cost functions

### ISSUE THREAD

**Issue by JzHuai0108** [ceres-solver/ceres-solver#998]
I encountered this problem again, and could not find a clue in one hour after perusing my code.
This is not an issue for ceres solver, and I am sure there is something wrong in my coding.
However, I'd like to put the issue here for those who encounter this issue in the future including myself.

Here is the log
```
I0823 09:00:40.941304 1045289 odometry.h:367] Ceres Solver Report: Iterations: 1, Initial cost: 1.155486e+10, Final cost: 1.155486e+10, Termination: NO_CONVERGENCE
```

I have written several optimization problems with the DynamicAutoDiffCostFunction and custom manifold for local parameterization.
As I remember, at least 4 of them work perfectly, converging to good results.

I have helped a colleague debugging the same above NO_CONVERGENCE issue.
Though we fixed the problem for his case, I was not sure which was the root cause. We made several changes.
a. For each cost term, we ensure only the used parameters are passed to the DynamicAutoDiffCostFunction.
b. We used the ceres::sqrt in evaluating the cost, instead of a sqrt which we couldn't trace its namespace (definitely not std, because std::sqrt does not work with ceres jets).
c. In the debugging, I make sure that the cost term changes by perturbing the parameters to the cost term.

Now for the new case, I have not found a clue about the reason why it does not converge. I will setup several tests to debug it.

**Comment by sandwichmaker** [comment]
try running the solver with -logtostderr -v=3 and see what the log says.

**Comment by JzHuai0108** [comment]
The logging options definitely help revealing the progress of ceres solver.

My debugging steps are
I setup a test problem. The total cost was huge like 1e40. 
Then I printed out the cost for each term, and found that the costs for the autodiff functions were huge and increasing.
Then I looked at the operator() function for the autodiff function.
I did not set the residuals to zero before accumulating the costs.
After fixing this error, the ceres solver's solution looks promising.
Of course, the cost term needs further tuning for good results.
